src/Uno.Toolkit.UI/Controls/SafeArea/SafeArea.cs

```diff
@@ -0,0 +1,541 @@
+		/// </summary>
+		private static bool AreBoundsAspectRatiosConsistent
+		{
+			get
+			{
+#if WINUI
+				return true;
+#else
+#if HAS_UNO
+				return ApplicationView.GetForCurrentView().VisibleBounds.GetOrientation() == XamlWindow.Current?.Bounds.GetOrientation();
+#else
+				return true;
+#endif
+#endif
+			}
+		}
+
+		private static Thickness GetWindowInsets(Rect? safeAreaOverride = null, bool withSoftInput = false)
+		{
+#if WINUI
+			return new();
+#else
+			var visibleBounds = GetVisibleBounds(safeAreaOverride, withSoftInput);
+			var bounds = XamlWindow.Current?.Bounds ?? Rect.Empty;
+
+			var result = new Thickness
+			{
+				Left = visibleBounds.Left - bounds.Left,
+				Top = visibleBounds.Top - bounds.Top,
+				Right = bounds.Right - visibleBounds.Right,
+				Bottom = bounds.Bottom - visibleBounds.Bottom,
+			};
+#if HAS_UNO
+			if (_log.IsEnabled(LogLevel.Debug))
+			{
+				_log.LogDebug($"WindowInsets={result} bounds={bounds} visibleBounds={visibleBounds}");
+			}
+#endif
+
+			return result;
+#endif
+		}
+
+
+		private static Rect GetVisibleBounds(Rect? safeAreaOverride = null, bool withSoftInput = false)
+		{
+#if WINUI
+			return new();
+#else
+			var visibleBounds = safeAreaOverride.IsEmptyOrZero()
+				? ApplicationView.GetForCurrentView().VisibleBounds
+				: safeAreaOverride.GetValueOrDefault();
+
+			if (withSoftInput)
+			{
+				var inputRect = InputPane.GetForCurrentView()?.OccludedRect ?? Rect.Empty;
+				if (inputRect.Height > 0)
+				{ 
+					var heightDiff = Math.Max(0, visibleBounds.Bottom - inputRect.Top);
+					visibleBounds.Height -= heightDiff;
+				}
+			}
+
+			return visibleBounds;
+#endif
+		}
+
+		public class SafeAreaDetails
+		{
+			private static readonly ConditionalWeakTable<FrameworkElement, SafeAreaDetails> _instances =
+				new ConditionalWeakTable<FrameworkElement, SafeAreaDetails>();
+			private readonly WeakReference _owner;
+			private readonly TypedEventHandler<global::Windows.UI.ViewManagement.ApplicationView, object> _visibleBoundsChanged;
+			private Rect? _overriddenVisibleBounds;
+			private InsetMask _insetMask;
+			private InsetMode _insetMode = InsetMode.Padding;
+			private readonly Thickness _originalPadding;
+			private readonly Thickness _originalMargin;
+			private Thickness _appliedPadding = new Thickness(0);
+			private Thickness _appliedMargin = new Thickness(0);
+
+			internal SafeAreaDetails(FrameworkElement owner)
+			{
+				_owner = new WeakReference(owner);
+
+				_originalMargin = owner.Margin;
+				_originalPadding = PaddingHelper.GetPadding(owner);
+
+				_visibleBoundsChanged = (s2, e2) => UpdateInsets();
+#if __IOS__
+				// For iOS, it's required to react on SizeChanged to prevent weird alignment
+				// problems with Text using the LayoutManager (NSTextContainer).
+				// https://github.com/unoplatform/uno/issues/2836
+
+				owner.SizeChanged += (s, e) => UpdateInsets();
+#endif
+				owner.LayoutUpdated += (s, e) => UpdateInsets();
+				owner.Loaded += (s, e) =>
```

Comment:
What is the owner here? Is it the `SafeArea` control?

Reply:
It would be whatever control that you are attaching the  properties to. Like `<Border SafeArea.Insets="Top" />, owner would be Border. But if you're wrapping a control like: 

```xml
<SafeArea Insets="Top">
  <Border />
</SafeArea>
```

Owner would be the SafeArea... need to double check this, I think maybe SafeArea should just forward the property values to it's child and attach the properties to the Border. So it should always be Border as the owner, hmm confused now.